GVFS/GVFS.Common/NamedPipes/QueuedPipeWriter.cs

```diff
@@ -0,0 +1,209 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.IO.Pipes;
+using System.Threading;
+
+namespace GVFS.Common.NamedPipes
+{
+    public class QueuedPipeWriter : IDisposable
+    {
+        private static readonly TimeSpan PollingInterval = TimeSpan.FromSeconds(30);
+
+        private readonly Func<NamedPipeClientStream> pipeFactory;
+        private readonly ConcurrentQueue<byte[]> queue = new ConcurrentQueue<byte[]>();
+        private readonly AutoResetEvent queuePulse = new AutoResetEvent(initialState: false);
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private bool isRunning;
+        private bool isDisposed;
+
+        public QueuedPipeWriter(Func<NamedPipeClientStream> pipeFactory)
+        {
+            this.pipeFactory = pipeFactory;
+        }
+
+        public event EventHandler<QueuedPipeWriterStateEventArgs> StateChanged;
+
+        public enum QueuedPipeWriterState
+        {
+            Stopped,
+            Failing,
+            Healthy,
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeWriter));
+            }
+
+            if (this.isRunning)
+            {
+                return;
+            }
+
+            this.isRunning = true;
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public void Enqueue(byte[] data)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                throw new InvalidOperationException($"{nameof(QueuedPipeWriter)} is not started");
+            }
+
+            this.queue.Enqueue(data);
+
+            // Wake the queue draining thread if it is waiting
+            this.queuePulse.Set();
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeWriter));
+            }
+
+            if (!this.isRunning)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            // Setting isRunning to false will additionally prevent any new messages from being enqueued.
+            this.isRunning = false;
```

Comment:
> This looks like it will throw a `InvalidOperationException` if any messages are enqueued while waiting to stop right? Is that what we want to do?

Do you think we should just drop the messages instead (`return;`) rather than throw?